contracts/mixins/ProtocolTokenHandler.sol

```diff
@@ -51,51 +65,75 @@ contract ProtocolTokenHandler is ProtocolTokenUser {
      */
     function deposit(uint256 _amount) public {
         require(_amount > 0, "amount should larger than zero");
 
         IERC20(protocolTokenAddress).safeTransferFrom(msg.sender, address(this), _amount);
-        protocolTokenHeld = protocolTokenHeld.add(_amount);
 
         emit Deposit(msg.sender, _amount);
     }
 
     /**
-     * @param _data the message/data from users to be signed by authorized signers, should include in order: 
-     * address of recipient, addresss of destination contract (should be this contract address), amount of tokens to be withdrawed, nounce
      * @param _signs array includes signs from authorized signers, amount/length should equal to requiredCount
+     * @param _recipient address of recipient
+     * @param _contractAddress addresss of destination contract (should be this contract address)
+     * @param _amount amount of tokens to be withdrawed
+     * @param _nonce nonce to avoid replay attacks
+     * @param _signs array includes signs from authorized signers
      */
-    function withdraw(bytes memory _data, bytes[] memory _signs) public {
-        _signs = new bytes[](requiredCount);
-        bytes32 _hash = keccak256(_data).toEthSignedMessageHash();
+    function withdraw(address _recipient, address _contractAddress, uint256 _amount, uint256 _nonce, bytes[] memory _signs) public {
+        require(_contractAddress == address(this), "contract address disagreement");
+        require(_amount > 0, "amount should larger than zero");
 
-        for (uint256 i = 0; i < requiredCount; i++) {
-            address signer = _hash.recover(_signs[i]);
-            require(isSigner[signer], "signer not authorized");
-            require(!verifiedSigners[signer], "signer verified");
-            verifiedSigners[signer] = true;
-        }
+        uint256 oldNonce = userNonce[msg.sender];
+        require(_nonce > oldNonce, "nonce smaller than last know nonce");
+        userNonce[msg.sender] = _nonce;
+
+        bytes32 _hash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(abi.encodePacked(_recipient, _contractAddress, _amount, _nonce))));
+        
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        address signer;
+        bytes memory sign;
+        bytes[] memory signs = new bytes[](requiredCount);
+        address[] memory verifiedSigners = new address[](requiredCount);
+
+        signs = _signs;
 
         for (uint256 i = 0; i < requiredCount; i++) {
-            verifiedSigners[_hash.recover(_signs[i])] = false;
-        }
+            sign = signs[i];
+            require(sign.length == 65, "length of sign is wrong");
 
-        (address recipient, address contractAddress, uint256 amount, uint256 nonce) = abi.decode(_data, (address, address, uint256, uint256));
+            assembly {
+                r := mload(add(sign, 0x20))
+                s := mload(add(sign, 0x40))
+                v := add(0x1b, byte(0, mload(add(sign, 0x60))))
+            }
```

Comment:
that's a way to do it and i guess figuring that out was not easy. i'm glad you learned how to access calldata using assembly, but you could have made your life easier and just pass  uint8[] v, bytes32[] r and bytes32[] s.

This is just a comment to let you know this would have been an option. you can leave it as it is ;)